concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,184 @@
+//! memory leaks as perfectly fine and even preferable since the need for
+//! cleanup and advanced allocation schemes is removed. The bump allocator here
+//! has two extra features:
+//!  - It keeps track of the number of active allocations in the `allocations`
+//!    field and resets the `next` pointer to `MIN_PTR_ADDR` if there are no
+//!    active allocations, thus allowing the reuse of memory.
+//!  - On deallocations it checks whether it is the very last block handed out
+//!    that is deallocated. If it is, then it moves the `next` pointer back to
+//!    the beginning of that block. Thus also allowing memory to be reused.
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    cell::UnsafeCell,
+    ptr,
+};
+
+/// The byte size of Wasm pages.
+const PAGE_SIZE: usize = 65536;
+/// FIXME: Handing out memory address 0 doesn't work, but starting with 1 does.
+const MIN_PTR_ADDR: usize = 1;
+
+/// A number of WebAssembly memory pages.
+#[derive(Eq, PartialEq, Copy, Clone)]
+struct PageCount(usize);
+
+impl PageCount {
+    /// The size in bytes. I.e. `PageCount * PAGE_SIZE`.
+    fn size_in_bytes(&self) -> usize { self.0 * PAGE_SIZE }
+}
+
+/// The number of pages returned from `memory_grow` to indicate out of memory.
+const ERROR_PAGE_COUNT: PageCount = PageCount(usize::MAX);
+
+/// This is an invalid implementation of Sync.
+/// The [BumpAllocator] cannot be safely used from multiple threads, but that is
+/// OK since it won't be in Wasm. This [Sync] implementation is required for
+/// defining it as the global allocator.
+unsafe impl Sync for BumpAllocator {}
+
+pub struct BumpAllocator {
+    /// The next pointer to be handed out.
+    next:        UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to 1 if
+    /// there are no more active allocations.
+    allocations: UnsafeCell<usize>,
+    /// Stores the last address given out, and allows for a small improvement
+    /// over the always-increasing bump allocator. Namely that if an item
+    /// `a` is allocated and deallocated without any allocations in between,
+    /// then the memory of `a` is reused.
+    last_alloc:  UnsafeCell<usize>,
+}
+
+impl BumpAllocator {
+    /// Create a new [BumpAllocator].
+    ///
+    /// # Safety
+    /// - Can only be used in single-threaded environments.
+    pub const unsafe fn new() -> Self {
+        Self {
+            next:        UnsafeCell::new(MIN_PTR_ADDR),
+            heap_end:    UnsafeCell::new(0),
+            allocations: UnsafeCell::new(0),
+            // Must be set to the same initial address as `next`.
+            last_alloc:  UnsafeCell::new(MIN_PTR_ADDR),
+        }
+    }
+
+    /// Grow the memory by `delta` pages, each of size [PAGE_SIZE].
```

Comment:
Suggestion: You could mentioned that the function returns the previous `PageCount`.